hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerSet.java

```diff
@@ -236,14 +244,43 @@ public ContainerReportsProto getContainerReport() throws IOException {
       ContainerDeletionChoosingPolicy deletionPolicy)
       throws StorageContainerException {
     Map<Long, ContainerData> containerDataMap = containerMap.entrySet().stream()
-        .filter(e -> deletionPolicy.isValidContainerType(
-            e.getValue().getContainerType()))
-        .collect(Collectors.toMap(Map.Entry::getKey,
-            e -> e.getValue().getContainerData()));
+        .filter(e ->
+            deletionPolicy.isValidContainerType(e.getValue().getContainerType())
+                && isDeletionAllowed(e.getValue().getContainerData())).collect(
+            Collectors.toMap(Map.Entry::getKey,
+                e -> e.getValue().getContainerData()));
     return deletionPolicy
         .chooseContainerForBlockDeletion(count, containerDataMap);
   }
 
+  private boolean isDeletionAllowed(ContainerData containerData) {
+    if (containerData.isClosed()) {
+      if (writeChannel instanceof XceiverServerRatis) {
+        try {
+          XceiverServerRatis ratisServer = (XceiverServerRatis) writeChannel;
+          long minReplicatedIndex = ratisServer.getMinReplicatedIndex(PipelineID
+              .valueOf(UUID.fromString(containerData.getOriginPipelineId())));
+          long containerBCSID = containerData.getBlockCommitSequenceId();
+          if (minReplicatedIndex != 0 && minReplicatedIndex < containerBCSID) {
```

Comment:
There can be a corner case where minReplicatedIndex is 0 and pipeline is still active?

Reply:
Will update the code to return -1 in case getMinReplicatedIndex call fail.